### PR TITLE
feat: add flow types

### LINKED ISSUE #42: Export flow types
We use flow types a lot, and are in the process of writing flow types for `mongoist`. I intend to publish these in `flow-typed`, but am wondering if @saintedlama has any interest in including those types either in this package (separate from the code), or using flow in the implementation itself. It seems that typescript may be a more broadly supported tool these days, though.

I'll resolve this either way, but the result may be a new PR against [flow-typed](https://github.com/flow-typed/flow-typed).

### PR DESCRIPTION
**Pull request by skeggse**
These are incomplete, but cover common surface area.

Also restricts published files:

```patch
@@ -2,26 +2,13 @@
 npm notice
 npm notice 📦  mongoist@2.4.0
 npm notice === Tarball Contents ===
-npm notice 112B   .babelrc
-npm notice 209B   .editorconfig
-npm notice 11B    .eslintignore
-npm notice 452B   .eslintrc
-npm notice 67B    .jsbeautifyrc
 npm notice 771B   LICENSE
 npm notice 16.5kB index.js.flow
 npm notice 3.5kB  lib/bulk.js
-npm notice 5.5kB  test/bulk.js
 npm notice 5.6kB  lib/collection.js
-npm notice 9.8kB  test/collection.js
 npm notice 3.3kB  lib/cursor.js
-npm notice 8.1kB  test/cursor.js
 npm notice 6.6kB  lib/database.js
-npm notice 9.2kB  test/database.js
 npm notice 1.7kB  index.js
-npm notice 823B   test/index.js
-npm notice 1.4kB  package.json
+npm notice 1.5kB  package.json
 npm notice 7.5kB  CHANGELOG.md
-npm notice 3.2kB  CODE_OF_CONDUCT.md
-npm notice 2.1kB  CONTRIBUTING.md
 npm notice 15.8kB README.md
-npm notice 715B   .travis.yml
```

Fixes #42

### CODE REVIEW

File: index.js.flow [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+    projection?: Projection,
+    options?: QueryOptions
+  ): Promise<Cursor<Doc>>;
+  findOne<Doc>(query: Query, projection?: Projection, options?: QueryOptions): Promise<Doc>;
+  findAndModify(options: FindAndModifyParams): Promise<mixed | null>;
+  count(query: Query): Promise<number>;
+  distinct<FieldType>(field: string, query: Query): Promise<FieldType[]>;
+  insert: typeof Collection.prototype.insertOne & typeof Collection.prototype.insertMany;
+  insertOne<Doc>(doc: Doc, options?: InsertOptions): Promise<$Call<AddObjectIDField, Doc>>;
+  insertMany<Docs>(docs: Docs, options?: InsertOptions): Promise<$TupleMap<Docs, AddObjectIDField>>;
```
Comment (skeggse) [inline index.js.flow]: :sparkles:

File: index.js.flow [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+  constructor(db: Database_, name: string): Collection;
+
+  connect(): Promise<InternalCollection>;
+  find<Doc>(query: Query, projection?: Projection, options?: QueryOptions): Promise<Doc[]>;
+  findAsCursor<Doc>(
+    query: Query,
+    projection?: Projection,
+    options?: QueryOptions
+  ): Promise<Cursor<Doc>>;
+  findOne<Doc>(query: Query, projection?: Projection, options?: QueryOptions): Promise<Doc>;
```
[the remark this answers is not in the document]
Comment (shils) [reply index.js.flow]: Seems reasonable to me. The collection class itself could be parameterized by `Doc`, but that seems like it'd require more casting at individual call-sites.

File: index.js.flow [hunk 3]
```diff
@@ -393,10 +391,10 @@ type IndexOptions = {|
   partialFilterExpression?: { [string]: mixed, ... },
   collation?: CollationParam,
   session?: ClientSession,
-|};
+|}>;
 
 type AggregationPipeline = mixed[];
```
Comment (shils) [inline index.js.flow]: This should be reasonably typeable, at the stage name level at least. E.g.

```
type AggregationStageType = '$match' | '$project' | '$limit' ...

type AggregationStage = { [AggregationStageType]: mixed }

type AggregationPipeline = AggregationStage[];
```

File: index.js.flow [hunk 4]
```diff
@@ -453,53 +451,66 @@ export type Projection = {
   ...
 };
 
-type AggregationUpdate = (
+type AggregationUpdate = $ReadOnlyArray<
```
Comment (shils) [inline index.js.flow]: 🎉

File: index.js.flow [hunk 5]
```diff
@@ -453,53 +451,66 @@ export type Projection = {
...
+  $pullAll?: { [string]: mixed, ... },
+  $push?: { [string]: mixed, ... },
+  $rename?: { [string]: mixed, ... },
+  $set?: { [string]: mixed, ... },
+  // TODO: forbid this option if upsert is not true.
+  $setOnInsert?: { [string]: mixed, ... },
+  $unset?: { [string]: mixed, ... },
+|};
+
+type ReadOnlyUpdate = $ReadOnly<{
```
Comment (shils) [inline index.js.flow]: "ReadOnly" + "Update" could be a bit confusing out of context. Not an issue if the type isn't exposed to the user.

Comment (skeggse) [reply index.js.flow]: Yeah I hear that - I'll add a clarifying comment.

File: index.js.flow (hunk 3 above)
Comment (skeggse) [reply index.js.flow]: 06d000acc4dc8c1b54462f41a787821a53f57048

File: index.js.flow (hunk 5 above)
Comment (skeggse) [reply index.js.flow]: cd5181ecf93c002b89a3feb8e1c5466ca449d6ec

**Pull request merged**